ntp-proto/src/algorithm/kalman/mod.rs

```diff
@@ -0,0 +1,358 @@
+                std::process::exit(exitcode::SOFTWARE);
+            }
+        } else {
+            self.timedata.accumulated_steps += change;
+            if !self.config.panic_threshold.is_within(change)
+                || !self
+                    .config
+                    .accumulated_threshold
+                    .map(|v| change.abs() < v)
+                    .unwrap_or(true)
+            {
+                error!("Unusually large clock step suggested, please manually verify system clock and reference clock state and restart if appropriate.");
+                std::process::exit(exitcode::SOFTWARE);
+            }
+        }
+    }
+
+    fn steer_offset(&mut self, change: f64, freq_delta: f64) -> Option<NtpTimestamp> {
+        self.check_offset_steer(change);
+        if change.abs() > self.algo_config.jump_threshold {
+            // jump
+            self.clock
+                .step_clock(NtpDuration::from_seconds(change))
+                .unwrap();
+            for (state, _) in self.peers.values_mut() {
+                state.process_offset_steering(change)
+            }
+            info!("Jumped offset by {}ms", change * 1e3);
+            None
+        } else {
+            // start slew
+            let freq = self
+                .algo_config
+                .slew_max_frequency_offset
+                .min(change.abs() / self.algo_config.slew_min_duration);
+            let duration = NtpDuration::from_seconds(change.abs() / freq);
+            info!(
+                "Slewing by {}ms over {}s",
+                change * 1e3,
+                duration.to_seconds()
+            );
+            Some(self.change_desired_frequency(-freq * change.signum(), freq_delta) + duration)
+        }
+    }
+
+    fn change_desired_frequency(&mut self, new_freq: f64, freq_delta: f64) -> NtpTimestamp {
+        let change = self.desired_freq - new_freq + freq_delta;
+        self.desired_freq = new_freq;
+        self.steer_frequency(change)
+    }
+
+    fn steer_frequency(&mut self, change: f64) -> NtpTimestamp {
+        self.freq_offset = (1.0 + self.freq_offset) * (1.0 + change) - 1.0;
+        self.clock.set_frequency(self.freq_offset).unwrap();
+        let freq_update = self.clock.now().unwrap();
+        for (state, _) in self.peers.values_mut() {
+            state.process_frequency_steering(freq_update, change)
+        }
+        info!(
+            "Changed frequency, current steer {}ppm, desired freq {}ppm",
+            self.freq_offset * 1e6,
+            self.desired_freq * 1e6,
+        );
+        freq_update
+    }
+
+    fn update_desired_poll(&mut self) {
+        self.timedata.poll_interval = self
+            .peers
+            .values()
+            .map(|(state, _)| state.get_desired_poll(&self.config.poll_limits))
+            .min()
+            .unwrap_or(self.config.poll_limits.max);
```

Comment:
why does max make sense here? We try to get the minimal poll from all the peers, so then `min` would intuitively make more sense to me.

Reply:
max makes sense here exactly because we try to get the minimal poll from all the peers. Using max then guarantees that this will only ever get lower if there are more peers, not higher. (for the min operation, maximum is essentialy the identity element)